src/dos/drive_local.cpp

```diff
@@ -194,42 +194,39 @@ bool localDrive::FileUnlink(char * name) {
 	safe_strcpy(newname, basedir);
 	safe_strcat(newname, name);
 	CROSS_FILENAME(newname);
-	char *fullname = dirCache.GetExpandName(newname);
-	if (unlink(fullname)) {
-		//Unlink failed for some reason try finding it.
-		struct stat buffer;
-		if (stat(fullname,&buffer)) return false; // File not found.
-
-		FILE* file_writable = fopen_wrap(fullname,"rb+");
-		if (!file_writable) return false; //No acces ? ERROR MESSAGE NOT SET. FIXME ?
-		fclose(file_writable);
-
-		//File exists and can technically be deleted, nevertheless it failed.
-		//This means that the file is probably open by some process.
-		//See if We have it open.
-		bool found_file = false;
-		for (Bitu i = 0;i < DOS_FILES;i++) {
-			if (Files[i] && Files[i]->IsName(name)) {
-				Bitu max = DOS_FILES;
-				while (Files[i]->IsOpen() && max--) {
-					Files[i]->Close();
-					if (Files[i]->RemoveRef()<=0) break;
+	const char *fullname = dirCache.GetExpandName(newname);
+
+	// Can we remove the file without issue?
+	if (std::remove(fullname) == 0) {
+		dirCache.DeleteEntry(newname);
+		return true;
+	}
+
+	// Maybe we've got it open, so scan our Files[] inventory and close it
+	bool found_internally = false;
+	for (size_t i = 0; i < DOS_FILES; ++i) {
+		if (Files[i] && Files[i]->IsName(name)) {
+			size_t max = DOS_FILES;
+			while (Files[i]->IsOpen() && --max) {
+				Files[i]->Close();
+				if (Files[i]->RemoveRef() <= 0) {
+					break;
 				}
-				found_file=true;
 			}
+			found_internally = true;
 		}
-		if (!found_file) return false;
+	}
```

Comment:
This was a great suggestion, because it made me check if there's anywhere else we can leverage the same Files[] loop.

Turns out both these functions operate similarly:
- localDrive::FileOpen
- localDrive::FileUnlink

Both of them loop through `Files[]` for a match on the incoming `filename`. In the case of `Open`, they simply `->Flush()` the handle, where as `Unlink` `->Close()` the handle.

But checkou tthe scanning code in `Open`, which Coverity is perfectly fine with:

``` c++
bool localDrive::FileOpen(DOS_File** file, char * name, Bit32u flags) {

        <snip>

	Bit8u i,drive=DOS_DRIVES;
	localFile *lfp = nullptr;
	for (i = 0; i < DOS_DRIVES; i++) {
		if (Drives[i] == this) {
			drive = i;
			break;
		}
	}

	for (i = 0; i < DOS_FILES; i++) {
		if (Files[i] && Files[i]->IsOpen() && Files[i]->GetDrive()==drive && Files[i]->IsName(name)) {
			lfp=dynamic_cast<localFile*>(Files[i]);
			if (lfp) lfp->Flush();
		}
	}
	FILE* fhandle = fopen(newname, type);
}
```

The big difference is `Open` double checks that the file exists on it's own "localDrive" drive `if (Drives[i] == this)`  and `Files[i]->GetDrive()==drive`.  But the localDrive implementation of `Unlink` doesn't include this check.

If we had more than one local drives, say C:, D:, and E: each holding same-named file (so the pool of Files[] contains all three same files), then if `Unlink()` was called for one of them then its loop could wind up forcing-closed `->Close()` either of the other open files on the other drives. So my guess is, the `Open` implementation is more correct.

Curious what you think.  If you agree, I'll implement a scanning function that can be used in both places.

Reply:
Ran some tests - indeed, the loop compares only the filename against Files[] that exist on other mounted drives too - here I'm just printing the drive letter as a number, but we see it's comparing against CON and LPT1 which are way beyond Z:. 

``` text
FS: comparing $$$DN$$$.LST against 2:/DN.OVR
FS: comparing $DN166$.BAT against 255:/CON
FS: comparing $DN166$.BAT against 255:/CON
FS: comparing $DN166$.BAT against 255:/LPT1
FS: comparing $DN166$.BAT against 2:/DN.OVR
FS: comparing $DN166$.MNU against 255:/CON
FS: comparing $DN166$.MNU against 255:/CON
FS: comparing $DN166$.MNU against 255:/LPT1
FS: comparing $DN166$.MNU against 2:/DN.OVR
FS: comparing $$$DN$$.LST against 255:/CON
FS: comparing $$$DN$$.LST against 255:/CON
FS: comparing $$$DN$$.LST against 255:/LPT1
FS: comparing $$$DN$$.LST against 2:/DN.OVR
FS: comparing $$$DN$$$.LST against 255:/CON
FS: comparing $$$DN$$$.LST against 255:/CON
FS: comparing $$$DN$$$.LST against 255:/LPT1
FS: comparing $$$DN$$$.LST against 2:/DN.OVR
```

Reply:
I've implemented and tested the delete using the new `FindOpenFile` function, and it's doing the right thing, and reduces code side in both the Open and Unlink functions.

Reply:
The function is added in the next commit to evolve this code commit-by-commit, so it can easily be bisected if there are any more very subtle issues (like the reinterpret cast).